### DISCUSSION TITLE
Where can I see the alternate download URLs for mods?

### DISCUSSION THREAD

**Discussion by ProgrammerFailure** [KSP-CKAN/CKAN#4074] [Q&A] [answered]
I was wondering where alternate download urls are stored. Seeing the "Preferred host" setting in CKAN, its pretty obvious one mod can have several download urls, but I could not find a way to see where said urls are stored.

**Comment by HebaruSan** [comment, marked as answer]
Hi @ProgrammerFailure, thanks for your question. There's not an easy way right now as this is fairly new functionality and we were mostly focused on trying to make it actually _work_, but it will most likely be enhanced in the future. Almost all of the mods with multiple download hosts are for KSP2, because we wanted to be careful about disrupting KSP1 users in case there was a problem with it.

You can see the full list of downloads this way:

1. Open this page: <http://status.ksp-ckan.space/>
2. Use the search box and checkboxes in the upper right to find the mod of interest
3. Hover the mod's row so the extra links appear under its identifier
4. Click the "metadata" link; a list of .ckan files will appear
5. Choose a specific version of that mod; the newest one is usually at the bottom
6. Find the `download` property; if the mod has multiple download hosts, this will be an array with multiple links like this:
   <https://github.com/KSP-CKAN/KSP2-CKAN-meta/blob/95fa482c3634da70a1376bbdc8010bc3aa26d4cd/FlightPlan/FlightPlan-0.9.1.ckan#L42-L45>
   ```json
       "download": [
           "https://github.com/schlosrat/FlightPlan/releases/download/0.9.1/flight_plan_v0.9.1.zip",
           "https://spacedock.info/mod/3359/Flight%20Plan/download/0.9.1"
       ],
   ```
   But if it's a single string, then that mod only has one download host, like this:
   <https://github.com/KSP-CKAN/KSP2-CKAN-meta/blob/main/HeadUpDisplay/HeadUpDisplay-0.8.0.ckan#L33>
   ```json
       "download": "https://spacedock.info/mod/3334/HUD%20:%20Head%20Up%20Display/download/0.8.0",
   ```

**Reply by ProgrammerFailure** [reply]
@HebaruSan 
Thanks for the clarification, the NetKAN status website answers most of my questions.

As a side note, can CKAN download from the repository directly? Since basically all mods have a github, it should be possible to download from the releases page, even if that isn't listed as a download location right? That would mostly eliminate spacedock.info for me, which is helpful because I have very slow download speeds from spacedock.

**Reply by HebaruSan** [reply]
Hi @ProgrammerFailure, good question! Indeed, that's exactly why we added the multi-host feature, to (eventually) make as many mods as possible downloadable from both GitHub and SpaceDock.

But if you're asking whether it's possible to download releases from GitHub _automatically_ for mods that are currently indexed from SpaceDock, the answer to that is No. Manual review is required due to inconsistencies in how mods are packaged and versioned. Someone has to look at the mod and say, "OK, this `v` prefix on the version strings on SpaceDock is missing on GitHub (or vice versa), so here's how we'll massage that part of the metadata," for example.

**Reply by ProgrammerFailure** [reply]
@HebaruSan 
Ok, looking forward to auto-github-downloads when implemented, this answers most of my questions. If CKAN metadata was updated so that github repo links went to a specific release (i.e, instead of repository = github.com/KSTS, it would be repo = github.com/KSTS/releases/version), then auto downloads would work right? 

I suggest that repo links in CKAN metadata support linking to a release on github, which would allow the repo link to act as a download link automatically. Then mod authors can just set repo link to a release for new versions, with no change on older versions. 

Also are we allowed to submit metadata changes for other people's mods? I would like to add github downloads to my most frequently downloaded mods, because I frequently reinstall everything.

**Reply by HebaruSan** [reply]
> If CKAN metadata was updated so that github repo links went to a specific release (i.e, instead of repository = github.com/KSTS, it would be repo = github.com/KSTS/releases/version), then auto downloads would work right?

No, if I understand what you're suggesting, that's a misunderstanding of how a variety of things work and wouldn't address any of the underlying problems. First, the "repository" link has nothing to do with how mod versions are found; they're two separate mechanisms. You could change those all you wanted and not affect anything. Second, manual maintenance of release-specific links would not be viable due to the scope and scale of changes constantly needed to keep up. Who would be in charge of monitoring all mods somehow to find the new releases if it worked that way? Third, finding a release isn't a problem; we do that already automatically with the current repo-specific links.

> I suggest that repo links in CKAN metadata support linking to a release on github, which would allow the repo link to act as a download link automatically. Then mod authors can just set repo link to a release for new versions, with no change on older versions.

Unfortunately, this wouldn't work. See above re:

> Manual review is required due to inconsistencies in how mods are packaged and versioned. Someone has to look at the mod and say, "OK, this v prefix on the version strings on SpaceDock is missing on GitHub (or vice versa), so here's how we'll massage that part of the metadata," for example.

**Reply by ProgrammerFailure** [reply]
@HebaruSan 
Ah, ok. Bit of a bummer. Thats all the questions I have, thanks for the support.
